Civi/Gpapi/ContractHelper/Adyen.php

```diff
@@ -2,150 +2,312 @@
+      'campaign_id'                             => $campaign_id,
       'check_permissions'                       => 0,
-      'membership_payment.membership_recurring_contribution' => $mandate['entity_id'],
-    );
-    $response = civicrm_api3('Contract', 'modify', $contract_modification);
+      'date'                                    => $start_date,
+      'id'                                      => $membership_id,
+      'medium_id'                               => $medium_id,
+      'membership_payment.cycle_day'            => $cycle_day,
+      'membership_payment.membership_annual'    => $annual_amount,
+      'membership_payment.membership_frequency' => $frequency,
+      'membership_type_id'                      => $membership_type_id,
+      'payment_method.adapter'                  => 'adyen',
+      'payment_method.cycle_day'                => $cycle_day,
+      'payment_method.payment_instrument_id'    => $payment_instrument_id,
+    ];
+
+    $response = civicrm_api3('Contract', 'modify', $modify_contract_params);
+
     civicrm_api3('Contract', 'process_scheduled_modifications', [
-      'id'                => $this->membershipId,
+      'id'                => $membership_id,
       'check_permissions' => 0,
     ]);
-    return reset($response['values'])['change_activity_id'];
+
+    $this->loadContract($membership_id);
   }
 
-  private function getCreditor($params) {
-    $fileFormat = civicrm_api3('OptionValue', 'getvalue', [
-      'return' => 'value',
-      'option_group_id' => 'sepa_file_format',
-      'name' => 'adyen',
-    ]);
-    $creditorLookup = [
-      'creditor_type' => 'PSP',
-      'sepa_file_format_id' => $fileFormat,
+  public function createInitialContribution(array $params) {
+    $psp_result_data = CRM_Utils_Array::value('psp_result_data', $params, []);
+    $merchant_reference = CRM_Utils_Array::value('merchantReference', $psp_result_data);
+    $psp_reference = CRM_Utils_Array::value('pspReference', $psp_result_data);
+
+    $create_order_params = [
+      'campaign_id'            => $this->recurringContribution['campaign_id'],
+      'contact_id'             => $params['contact_id'],
+      'contribution_recur_id'  => $this->recurringContribution['id'],
+      'contribution_status_id' => 'Pending',
+      'financial_type_id'      => $this->recurringContribution['financial_type_id'],
+      'invoice_id'             => $merchant_reference,
+      'payment_instrument_id'  => $this->recurringContribution['payment_instrument_id'],
+      'receive_date'           => $this->membership['join_date'],
+      'sequential'             => TRUE,
+      'source'                 => 'OSF',
+      'total_amount'           => $this->recurringContribution['amount'],
     ];
-    if (empty($params['currency'])) {
-      $config = \CRM_Core_Config::singleton();
-      $creditorLookup['currency'] = $config->defaultCurrency;
+
+    $order_result = civicrm_api3('Order', 'create', $create_order_params);
```

Comment:
For SEPA and PSPSEPA, when the initial contribution is created, `CRM_Sepacustom_InstallmentProcessor::installmentCreated()` will create a corresponding `MembershipPayment` entity (note: this is Core's `civicrm_membership_payment`, not `civicrm_value_membership_payment`, which conveniently has the same name) connecting the contribution with its membership. We don't trigger that here and it's a SEPA thing, so we don't want to couple to tightly to that, so I would suggest just doing that manually based on `$order_result['id']` and the membership ID.